refactor(user): refactor user to use constructor injection #656

acocinas opened the pull request:
✅ Summary
This Pull Request addresses Task #656 (subtask of user story #644), which required refactoring the User microservice to eliminate the use of @Autowired in favor of constructor injection. The analysis revealed that the ContentLengthFilter and MaxLengthURIFilter classes were using a deprecated form of dependency injection. This PR resolves both instances, ensuring the microservice adheres to modern best practices.

🛠️ Implemented Changes
Migration to Constructor Injection:
The @Autowired annotations have been removed from the ContentLengthFilter and MaxLengthURIFilter classes. All dependencies are now injected exclusively through the constructor, with final fields, promoting immutability and readability.

Addition of Unit Tests:
New unit tests have been added for both ContentLengthFilter and MaxLengthURIFilter to verify their correct functionality after the refactoring. The tests use Mockito to mock dependencies, ensuring they are reliable and focused.

Justifies the technical approach
The chosen approach was to eliminate all instances of @Autowired from the microservice's production code. This refactoring aligns with the "Definition of Done" for this project's technical debt. This design:

Improves code quality and testability, making the classes easier to unit test without needing a full Spring context.

Enhances readability by making all dependencies explicit and mandatory via the constructor.

Promotes immutability by allowing dependencies to be declared as final.

⚠️ Technical Debt & Justification
This PR successfully resolves a specific piece of technical debt by completing the refactoring for the User microservice. The remaining refactoring tasks for the Document microservice will be addressed in the next PR and Challenge microservices will be addressed in future sprints as part of the overarching user story #644.

📋 PR Author Self-check
[✅] The @Autowired annotations have been removed from all production code in the User microservice.
[✅] The ContentLengthFilter and MaxLengthURIFilter classes now use constructor injection with final fields.
[✅] New unit tests have been added for both refactored classes.
[✅] All related tests pass successfully, confirming no regressions.
[✅] This is PR #962 for user story #644 and task #656.

acocinas committed:
feat: Refactor MaxLengthURIFilter to use constructor injection

acocinas committed:
feat: Refactor ContentLengthFilter to use constructor injection

acocinas committed:
test: Add tests for User microservice filters

Heyvkram reviewed (approved):
This is an excellent refactoring PR!

I've reviewed the changes and the approach is solid. The move to constructor injection in both the ContentLengthFilter and MaxLengthURIFilter has been executed perfectly, and using final fields for the dependencies is a great touch for immutability.

The tests are well-written, cover both the success and failure scenarios for each filter, and significantly increase our confidence in this part of the codebase.

Overall, this PR successfully achieves its goal, improves the code quality, and aligns perfectly with our best practices.

No further changes are needed from my side. Approved!

isaac-diez commented on itachallenge-user/src/test/java/com/itachallenge/user/filter/MaxLengthURIFilterTest.java:
What if there is no URI or URI length is 0? Is it worth testing this scenario?

acocinas replied:
I've added the new test case to MaxLengthURIFilterTest to handle null or empty URIs. This uncovered a NullPointerException that wasn't previously tested, so thank you for helping me find that bug!

isaac-diez commented on itachallenge-user/src/main/java/com/itachallenge/user/filter/MaxLengthURIFilter.java:
I see that you have used explicit constructor injection in ContentLengthFilter() and @RequiredArgsConstructor in MaxLengthURIFilter(). Both seem fine, but is this consistent?

acocinas replied:
I've made the constructor injection consistent across both filters, migrating ContentLengthFilter to use @RequiredArgsConstructor for better readability.

isaac-diez reviewed (changes requested):
Great work, Alfonso. A very well done PR.
I have commented a couple of small things regarding consistency of constructor injection and a possible third test when the URI is missing.

acocinas committed:
feat: Fix and refactor User filters

This commit completes the refactoring of the user filters. It addresses two key issues:
 **Refactoring:** The @Autowired annotation has been removed from ContentLengthFilter to align with project standards.
 **Bug fix:** MaxLengthURIFilter.java now correctly handles null URIs, fixing a NullPointerException. The corresponding unit test has been updated to validate this fix.

acocinas commented:
> Great work, Alfonso. A very well done PR. I have commented a couple of small things regarding consistency of constructor injection and a possible third test when the URI is missing.

Hey Isaac,

Thanks so much for the detailed review. I really appreciate your comments; they were incredibly helpful in improving the code.

I have addressed all the changes you requested:

I've made the constructor injection consistent across both filters, migrating ContentLengthFilter to use @RequiredArgsConstructor for better readability.

I've added the new test case to MaxLengthURIFilterTest to handle null or empty URIs. This uncovered a NullPointerException that wasn't previously tested, so thank you for helping me find that bug!

The code is now more robust and aligned with our Definition of Done. The test suite is also more comprehensive.

Thanks again for the great feedback! The PR is ready for your review.

acocinas committed:
feat: Fix maintainability issues and refactor MaxLengthURIFilter

isaac-diez reviewed (approved):
Looks good to me ✅
Thanks for unifying the constructor injection and adding the test case for requests without a query string. The implementation is now consistent and well-covered. Approving this PR.

acocinas committed:
docs: Update changelog and .env for User microservice refactoring

acocinas committed:
complete changelog information with taiga and pr

ivilarop commented:
@acocinas @isaac-diez  

Is the taiga task missing from the PR name? Why did you miss this? Are you following the PR checklist?

Numbering like 1/2, 2/2 only makes sense when the PRs are dependent on each other, that is:

The second one can't be merged until the first one is in.

Or the second one breaks if the first one isn't merged first.

If in your case the PRs are independent (each one is mergeable and functional on its own, even if they're part of the same refactoring initiative), then there's no need to use numbering. Just clarifying this in the description is enough.

ivilarop commented on itachallenge-user/src/main/java/com/itachallenge/user/filter/MaxLengthURIFilter.java:
```diff
@@ -4,30 +4,32 @@
-                ((requestHttp.getQueryString()==null) ?
-                        0:requestHttp.getQueryString().length());
+
+        String requestURL = (requestHttp.getRequestURL() != null) ? requestHttp.getRequestURL().toString() : "";
```
In this PR, the goal was to refactor dependency injection by removing @Autowired.
During the review, the idea arose to add an additional test for MaxLengthURIFilter, which revealed a NullPointerException and required a change to the implementation.

👉 Do you think this new test (and the bug fix) should have been included in this same PR, even though it wasn't directly related to the refactor?

Or do you think it would have been more appropriate to open a separate PR to treat it as a bugfix or technical debt later, keeping this PR focused only on the refactor?

acocinas replied:
Decision Justification
Thank you for the opportunity to review the scope of this PR in more detail. The decision to include the bug fix in the same PR, even though the primary goal was refactoring, was made with the following approach in mind:

Efficiency and Context: The bug we found was a NullPointerException—a critical error with a very small fix. We felt that the most efficient way to proceed was to address it immediately, as the refactoring test had uncovered it. Creating a separate PR would have added a bureaucratic step for a minor change.

Code Quality: As a team, our goal is to maintain code quality. Merging a PR with a known bug in the code, even if it wasn't directly related to the refactoring, didn't seem like the best option. For this reason, we decided to include the solution, making the PR not just a refactor, but a code quality improvement.

ivilarop replied:
@acocinas 

Thanks for the explanation; what you're saying makes a lot of sense. 👍

As a team reflection, the important thing is to find the balance:

Efficiency and quality: Fixing a critical bug detected at the same time avoids merging broken code, and in this case, it was a small, justified fix.

Focus on PRs: In general, it's recommended that PRs maintain a clear and limited scope (refactor, feature, or bugfix). This facilitates review and the history of changes. If the bugfix had been larger or non-critical, it probably would have been cleaner to open a separate PR.

👉 The key is in the criteria:

Critical bug + small fix → can be included in the same PR.

Non-critical bug or larger fix → better to have a separate PR to avoid confusing objectives.

In this way, we not only improve code quality, but also the clarity of our review process. 🚀

you can solve the conversation

ivilarop commented on itachallenge-user/src/test/java/com/itachallenge/user/filter/ContentLengthFilterTest.java:
Same as the previous comment. Would it be outside the initial scope?

acocinas replied:
same as previous comment

ivilarop replied:
you can solve the conversation

ivilarop commented on itachallenge-user/src/test/java/com/itachallenge/user/filter/MaxLengthURIFilterTest.java:
Same as the previous comment. Would it be outside the initial scope?

acocinas replied:
same as the previous comment

ivilarop replied:
you can solve the conversation

acocinas commented:
> @acocinas @isaac-diez
> 
> Is the taiga task missing from the PR name? Why did you miss this? Are you following the PR checklist?
> 
> Numbering like 1/2, 2/2 only makes sense when the PRs are dependent on each other, that is:
> 
> The second one can't be merged until the first one is in.
> 
> Or the second one breaks if the first one isn't merged first.
> 
> If in your case the PRs are independent (each one is mergeable and functional on its own, even if they're part of the same refactoring initiative), then there's no need to use numbering. Just clarifying this in the description is enough.

Thanks for the review and for the clarifications. I really appreciate the feedback; it's very helpful for learning and improving our process.

You are absolutely right. I made a few mistakes that I have already corrected:

Taiga Task Number in PR Title: My apologies for missing the task number in the PR title. I have already updated the PR titles to follow the correct format, including the Taiga task number for proper tracking.

PR Numbering (1/2, 2/2): I now understand the distinction. The PRs for the user and document microservices are independent, as merging one does not affect the functionality of the other. I have removed the numbering from the titles and clarified their independence in the descriptions.

PR Checklist: I have reviewed the checklist and will ensure I follow all points to the letter to avoid these issues in the future.

My next step is to get the updated PRs re-reviewed.

Thanks again for the guidance.

ivilarop commented:
> > @acocinas @isaac-diez
> > Is the taiga task missing from the PR name? Why did you miss this? Are you following the PR checklist?
> > Numbering like 1/2, 2/2 only makes sense when the PRs are dependent on each other, that is:
> > The second one can't be merged until the first one is in.
> > Or the second one breaks if the first one isn't merged first.
> > If in your case the PRs are independent (each one is mergeable and functional on its own, even if they're part of the same refactoring initiative), then there's no need to use numbering. Just clarifying this in the description is enough.
> 
> Thanks for the review and for the clarifications. I really appreciate the feedback; it's very helpful for learning and improving our process.
> 
> You are absolutely right. I made a few mistakes that I have already corrected:
> 
> Taiga Task Number in PR Title: My apologies for missing the task number in the PR title. I have already updated the PR titles to follow the correct format, including the Taiga task number for proper tracking.
> 
> PR Numbering (1/2, 2/2): I now understand the distinction. The PRs for the user and document microservices are independent, as merging one does not affect the functionality of the other. I have removed the numbering from the titles and clarified their independence in the descriptions.
> 
> PR Checklist: I have reviewed the checklist and will ensure I follow all points to the letter to avoid these issues in the future.
> 
> My next step is to get the updated PRs re-reviewed.
> 
> Thanks again for the guidance.

Great job responding to feedback so quickly! 👏

You corrected the PR titles with Taiga's task number.

You clarified the independence of the PRs and removed the misleading numbering.

You confirmed that you now follow the PR checklist carefully.

This shows that you're learning quickly and improving your workflow. 🚀
